src/mapblock.h

```diff
@@ -450,60 +437,85 @@ class MapBlock
 	// clearObject and return removed objects count
 	u32 clearObjects();
 
+	static const u32 ystride = MAP_BLOCKSIZE;
+	static const u32 zstride = MAP_BLOCKSIZE * MAP_BLOCKSIZE;
+
+	static const u32 nodecount = MAP_BLOCKSIZE * MAP_BLOCKSIZE * MAP_BLOCKSIZE;
+
 private:
 	/*
 		Private methods
 	*/
 
 	void deSerialize_pre22(std::istream &is, u8 version, bool disk);
 
-public:
 	/*
-		Public member variables
-	*/
+	 * PLEASE NOTE: When adding something here be mindful of position and size
+	 * of member variables! This is also the reason for the weird public-private
+	 * interleaving.
+	 * If in doubt consult `pahole` to see the effects.
+	 */
 
+public:
 #ifndef SERVER // Only on client
 	MapBlockMesh *mesh = nullptr;
-#endif
-
-	NodeMetadataList m_node_metadata;
-	StaticObjectList m_static_objects;
 
-	static const u32 ystride = MAP_BLOCKSIZE;
-	static const u32 zstride = MAP_BLOCKSIZE * MAP_BLOCKSIZE;
-
-	static const u32 nodecount = MAP_BLOCKSIZE * MAP_BLOCKSIZE * MAP_BLOCKSIZE;
-
-	//// ABM optimizations ////
-	// Cache of content types
-	std::unordered_set<content_t> contents;
-	// True if content types are cached
-	bool contents_cached = false;
-	// True if we never want to cache content types for this block
-	bool do_not_cache_contents = false;
 	// marks the sides which are opaque: 00+Z-Z+Y-Y+X-X
-	u8 solid_sides {0};
+	u8 solid_sides = 0;
+#endif
 
 private:
-	/*
-		Private member variables
-	*/
+	// see isOrphan()
+	bool m_orphan = false;
 
-	// NOTE: Lots of things rely on this being the Map
-	Map *m_parent;
 	// Position in blocks on parent
 	v3s16 m_pos;
 
-	/* This is the precalculated m_pos_relative value
-	* This caches the value, improving performance by removing 3 s16 multiplications
-	* at runtime on each getPosRelative call
-	* For a 5 minutes runtime with valgrind this removes 3 * 19M s16 multiplications
-	* The gain can be estimated in Release Build to 3 * 100M multiply operations for 5 mins
-	*/
+	/* Precalculated m_pos_relative value
+	 * This caches the value, improving performance by removing 3 s16 multiplications
+	 * at runtime on each getPosRelative call.
+	 * For a 5 minutes runtime with valgrind this removes 3 * 19M s16 multiplications.
+	 * The gain can be estimated in Release Build to 3 * 100M multiply operations for 5 mins.
+	 */
 	v3s16 m_pos_relative;
 
+	/*
+		Reference count; currently used for determining if this block is in
+		the list of blocks to be drawn.
+	*/
+	short m_refcount = 0;
+
+	/*
+	 * Note that this is not an inline array because that has implications for
+	 * heap fragmentation (the array is exactly 16K), CPU caches and/or
+	 * optimizability of algorithms working on this array.
+	 */
+	MapNode *const data; // of `nodecount` elements
```

Comment:
Could tell the compiler that it's aligned:
```suggestion
	alignas(__STDCPP_DEFAULT_NEW_ALIGNMENT__) MapNode *const data; // of `nodecount` elements
```
For `__STDCPP_DEFAULT_NEW_ALIGNMENT__`, see: https://en.cppreference.com/w/cpp/preprocessor/replace

(Same for `getData()`.)

Reply:
`MapNode` has `alignas(u32)` so maybe it already knows.

Reply:
The u32 alignment is already there in master. But if it's allocated with `new`, it has even more alignment. (`__STDCPP_DEFAULT_NEW_ALIGNMENT__` seems to be 16 on gcc and clang.)